torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -238,10 +266,48 @@ def to_sharded(self) -> None:
         self.free_all_gather_output()
         self.sharded_state = ShardedState.SHARDED
 
+    def to_sharded_post_forward(self) -> None:
+        if self.is_dtensor:
+            raise NotImplementedError(
+                "Resharding to smaller mesh with TP is not supported yet"
+            )
+        self._assert_in_states(ShardedState.UNSHARDED)
+        assert self.post_forward_mesh_info is not None  # mypy
+        shard_world_size = self.post_forward_mesh_info.shard_mesh_size
+        if (numel := self.all_gather_output.numel()) % shard_world_size != 0:
+            _raise_assert_with_print(
+                f"All-gather output size ({numel}) must be divisible by the shard "
+                f"world size ({shard_world_size})"
+            )
+        shard_rank = self.post_forward_mesh_info.shard_mesh_rank
+        sharded_numel = numel // shard_world_size
+        self._sharded_post_forward_param_data = (
+            self.all_gather_output.narrow(0, sharded_numel * shard_rank, sharded_numel)
+        ).clone()  # clone to be able to free all-gather output
+        sharded_post_forward_tensor = torch.as_strided(
+            self._sharded_post_forward_param_data,
```

Comment:
any particular reason we need this as strided call? iirc all model params should be contiguous and therefore the `_sharded_post_forward_param_data` should also be contiguous?

Reply:
The `as_strided` is for doing a view from 1D -> ND _and_ remove the padding in one call. `self._sharded_post_forward_param_data` is 1D since we narrowed into `self.all_gather_output`, which was 1D.